crates/typst-cli/src/update.rs

```diff
@@ -0,0 +1,311 @@
+use std::{
+    env, fs,
+    io::{self, BufReader, Cursor, ErrorKind, Read, Seek},
+    path::Path,
+};
+
+use semver::Version;
+use serde::Deserialize;
+use typst::{diag::bail, diag::StrResult, eval::eco_format};
+use xz2::bufread::XzDecoder;
+use zip::ZipArchive;
+
+use crate::args::UpdateCommand;
+
+const TYPST_GITHUB_ORG: &str = "typst";
+const TYPST_REPO: &str = "typst";
+
+/// Self update the typst CLI binary.
+///
+/// Fetches a target release or the latest release (if no version was specified)
+/// from GitHub, unpacks it and self replaces the current binary with the
+/// pre-compiled asset from the downloaded release.
+pub fn update(command: UpdateCommand) -> StrResult<()> {
+    match self_update_permitted()? {
+        SelfUpdatePermission::Deny => {
+            eprintln!("Self-update is disabled for this build of the typst cli");
+            eprintln!("You should probably use your system package manager");
+            bail!("update failed");
+        }
+        SelfUpdatePermission::Permit => {}
+    }
+
+    if let Some(ref version) = command.version {
+        let current_tag = env!("CARGO_PKG_VERSION").parse().unwrap();
+        if !command.force && version < &current_tag {
+            eprintln!("Certain downgraded Typst versions will not have the update command available");
+            eprintln!("Forcing a downgrade might break your installation");
+            eprintln!(
+                "You can force a downgrade by running `typst update <VERSION> --force`"
+            );
+            bail!("update failed");
+        }
+    }
+
+    let current_exe = env::current_exe()
+        .map_err(|err| eco_format!("failed to grab current exe path: {}", err))?;
+
+    #[cfg(target_os = "linux")]
+    let root_backup_dir = dirs::state_dir()
+        .or_else(|| dirs::data_dir())
+        .expect("unable to locate local data or state directories");
+    #[cfg(not(target_os = "linux"))]
+    let root_backup_dir =
+        dirs::data_dir().expect("unable to locate local data directory");
+    let backup_dir = root_backup_dir.join("typst");
+    let backup = backup_dir.join("typst_backup.part");
```

Comment:
this could be extracted into a function

Reply:
Now in `fn backup_path() -> StrResult<PathBuf>`.